include/arch/arm/arch/machine.h

```diff
@@ -74,12 +74,12 @@ static inline void Arch_finaliseInterrupt(void)
 }
 
 /* Update the value of the actual regsiter to hold the expected value */
-static inline exception_t Arch_setTLSRegister(word_t tls_base)
+static inline bool_t Arch_setTLSRegister(word_t tls_base)
```

Comment:
Replacing `exception_t` with `bool_t` is not a good idea. We have a whole framework set up for dealing with `exception_t` in the proofs and we'd have to duplicate it for `bool_t` for no real gain.

If the function can never return an exception, then `void` is fine, we can adjust for that in the proofs. If it can fail, then `exception_t` is the right one to use. `bool_t` is not a suitable replacement.

To understand this structure, it's best to look at the corresponding Haskell code: the types there carry more information. There are three main types of exceptions used in this (errors, faults, preemption). 

The Haskell spec is [here](https://github.com/seL4/l4v/blob/master/spec/haskell/), the main syscall structure is in [this file](https://github.com/seL4/l4v/blob/master/spec/haskell/src/SEL4/API/Syscall.lhs). The `Kernel` monad is normal execution, `KernelP` is execution that may be preempted, faults and (user-level) errors are `KernelF`.

To make sure that a function can safely change type in C, you would change the corresponding type in Haskell and propagate that up to the call sites. If this type checks, the change is safe, and you can also do it in C. Using something like `Bool` there would destroy that check.

Reply:
Thanks, I will dig into the Haskell code a bit to better understand this. I've kept the `exception_t` return type now, as this does not make a difference and the compiler should be able to optimize this anyway. However, this is code in `handleUnknownSyscall()` that is active only for `CONFIG_SET_TLS_BASE_SELF`. So I wonder, it this TLS register modification more then just a development feature then?